Add stats table

chowington opened the pull request.

chowington commented:
Feedback requested:

- Using `<table>`
- The method I implemented of using either the `MosaicData` or `TwoByTwoData` type in the mosaic viz component--maybe better to use `io-ts` for this?
- Table formatting

dmfalke commented on src/lib/core/components/visualizations/implementations/MosaicVisualization.tsx:
```diff
@@ -240,15 +248,22 @@ function MosaicViz(props: Props) {
+      if (isTwoByTwo) {
+        const response = dataClient.getTwoByTwo(
+          computation.type,
+          params as MosaicRequestParams
```
I don't think this is necessary
```suggestion
          params
```

chowington replied:
Thanks Dave! Resolved.

dmfalke commented on src/lib/core/components/visualizations/implementations/MosaicVisualization.tsx:
```diff
@@ -240,15 +248,22 @@ function MosaicViz(props: Props) {
+      } else {
+        const response = dataClient.getMosaic(
+          computation.type,
+          params as MosaicRequestParams
```
I don't think this is necessary
```suggestion
          params
```

chowington replied:
Resolved

dmfalke reviewed (changes requested):
Hi @chowington , this is looking good. There are some unnecessary type assertions. They should be removed to prevent issues in the future.

dmfalke commented:
> Feedback requested:
> 
>     * Using `<table>`

I think using a table element is fine.

> 
>     * The method I implemented of using either the `MosaicData` or `TwoByTwoData` type in the mosaic viz component--maybe better to use `io-ts` for this?

I don't understand this.

> 
>     * Table formatting

I would add some padding between table cells and lighten the background a bit (maybe `#eaeaea`).

Can you also add this to the RXC viz?

bobular commented on src/lib/core/api/data-api.ts:
```diff
@@ -321,6 +321,44 @@ export const MosaicResponse = type({
   }),
 });
 
+export type TwoByTwoResponse = TypeOf<typeof TwoByTwoResponse>;
```
I think you could reuse MosaicResponse here and union in the extra bits needed - especially since both responses start with the `mosaic` key.

Also it looks like conttable/MosaicResponse should include validation for  `sampleSizeTable` ?

chowington replied:
Thanks Bob! Updated to reuse MosaicResponse using io-ts `intersection` (not `union`, confusingly).

`sampleSizeTable` can be in a generic MosaicResponse, yes.

bobular commented on src/lib/core/components/visualizations/implementations/MosaicVisualization.tsx:
```diff
@@ -240,15 +248,22 @@ function MosaicViz(props: Props) {
-        ? twoByTwoResponseToData(await response)
-        : mosaicResponseToData(await response);
+
+      if (isTwoByTwo) {
```
Regarding your question
> The method I implemented of using either the MosaicData or TwoByTwoData type in the mosaic viz component--maybe better to use io-ts for this?

Is this where you mean?  I think it's fine to make two different requests and have data-api.ts validate them with two different validators...

chowington replied:
Yes, here and the above. Just seems like I'm doing the "either" thing io-ts is good at

dmfalke replied:
I agree with @bobular. This looks fine to me, as well.

bobular reviewed (commented):
I have some other comments I can't make here, and doesn't just apply to this visualization - I'll add a new issue about it.

chowington commented on src/lib/core/components/visualizations/implementations/MosaicVisualization.tsx:
```diff
@@ -203,7 +210,7 @@ function MosaicViz(props: Props) {
 
   const data = usePromise(
-    useCallback(async (): Promise<MosaicData> => {
+    useCallback(async (): Promise<MosaicData | TwoByTwoData> => {
```
> The method I implemented of using either the MosaicData or TwoByTwoData type in the mosaic viz component--maybe better to use io-ts for this?

Here

chowington commented:
> 
> 
> > Feedback requested:
> > ```
> > * Using `<table>`
> > ```
> 
> I think using a table element is fine.
> 
> > ```
> > * The method I implemented of using either the `MosaicData` or `TwoByTwoData` type in the mosaic viz component--maybe better to use `io-ts` for this?
> > ```
> 
> I don't understand this.
> 
> > ```
> > * Table formatting
> > ```
> 
> I would add some padding between table cells and lighten the background a bit (maybe `#eaeaea`).
> 
> Can you also add this to the RXC viz?

Thanks Dave! I added comments to where I was referencing. I can work on the stats table for the RxC, yeah.

dmfalke reviewed (approved):
Thanks, @chowington. These changes look good.
